### PR TITLE
Tighten token permissions in GitHub actions

### PR DESCRIPTION
**Pull request by dra27**
Tighter alternative to #11361. Tested on my fork, so opening a PR to ocaml/ocaml adds an additional test. In theory, this PR:

- Removes all privileges from the `GITHUB_TOKEN` in the GitHub Actions runners
- Disables `git` commands automatically using it in any subsequent (`actions/checkout@v2` does this)
- Tweaks the Changes check so that PRs which didn't have a `Changes` entry and then had the label added afterwards don't report "x - 1 of x checks passed" when merging

Closes: #11361

### CODE REVIEW

**Comment by dra27** [comment]
- Changes check failing [log entry](https://github.com/ocaml/ocaml/runs/7217355539?check_suite_focus=true#step:4:8)
- Changes check skipping [log entry](https://github.com/ocaml/ocaml/runs/7217378463?check_suite_focus=true#step:4:8) (after label added)

Curiously, the [Documentation](https://docs.github.com/en/rest/overview/permissions-required-for-github-apps#permission-on-issues) suggests that the API call should require `issues: read`, but it doesn't seem to 🤷‍♂️

File: tools/ci/actions/check-changes-modified.sh
Comment (xavierleroy) [inline tools/ci/actions/check-changes-modified.sh]: Perhaps this comment needs updating?

File: tools/ci/actions/check-changes-modified.sh
Comment (xavierleroy) [inline tools/ci/actions/check-changes-modified.sh]: Looks like black magic, but if you're confident...

**Review by xavierleroy — APPROVED** [review]
Looks good to me, thanks, but my look is quite superficial.  Would be nice to have a second opinion by someone who really knows Github actions.  @avsm perhaps?

File: .github/workflows/build.yml
Comment (xavierleroy) [inline .github/workflows/build.yml]: (Forgot to save this comment.)  Quoting from the doc:

> If you specify the access for any of these scopes, all of those that are not specified are set to none

So, the two lines above actually set all permissions to `none`.  That's on the safe side, great!  But that's not explicit at all...  Quoting the doc again:

> You can use the following syntax to disable permissions for all of the available scopes: `permissions: {}`

Could this be a clearer way to express what is meant here?

Comment (dra27) [reply .github/workflows/build.yml]: It does indeed - I've used that and added a comment explaining why it's there

File: tools/ci/actions/check-changes-modified.sh
Comment (dra27) [reply tools/ci/actions/check-changes-modified.sh]: It's a combination of a rune from the [GitHub Actions docs](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#example-2-calling-the-rest-api) and the original rune (due to @gasche) from the old Travis scripts, seasoned with a little bashism 🙂

File: .github/workflows/build.yml
Comment (dra27) [inline .github/workflows/build.yml:51]: This line is no longer necessary with the `persist-credentials` change above (and, it turns out, `git` exits with code 5 if you unset a configuration value which isn't set)

File: tools/ci/actions/check-changes-modified.sh
Comment (xavierleroy) [reply tools/ci/actions/check-changes-modified.sh]: What about
```
elif curl .... | grep -q 'no-change-entry-needed'; then
```
(No need to store the output of grep, then check that it's nonempty.)

**Comment by xavierleroy** [comment]
This PR is sinking into oblivion, like many others by @dra27.  I don't insist on cleaning up the shell scripting, but I would still very much like a review by someone who understands Github's infrastructure and permission model, if this exists.

File: tools/ci/actions/check-changes-modified.sh
Comment (dra27) [reply tools/ci/actions/check-changes-modified.sh]: Indeed - amended

**Comment by xavierleroy** [comment]
Two months later, the PR was improved (thanks @dra27) but is still waiting for a review by someone who understands Github's infrastructure and permission model, if this exists.

**Comment by DemiMarie** [comment]
> Two months later, the PR was improved (thanks @dra27) but is still waiting for a review by someone who understands Github's infrastructure and permission model, if this exists.

Looks fine to me as far as I can tell, though I don’t know anything more than you about this.  I would merge it and see if anything breaks, which it probably won’t.

**Comment by xavierleroy** [comment]
Thanks for the second opinion.  It's well known that security fixes receive the least amount of reviewing among all contributions :-(  I'll merge and we'll see what happens.

**Comment by DemiMarie** [comment]
> Thanks for the second opinion. It's well known that security fixes receive the least amount of reviewing among all contributions :-( I'll merge and we'll see what happens.

You’re welcome.  In this particular case, another factor was that none of this makes it into a relase, so it can (unless I have missed something) trivially be reverted if needed.

**Pull request merged**
